src/mococrw/mac.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+#include <memory>
+#include <vector>
+#include "openssl_wrap.h"
+
+namespace mococrw
+{
+
+/**
+ * @brief The MessageAuthenticationCode class is the abstract base class for the different implementation of
+ * message authentication code.
+ */
+class MessageAuthenticationCode {
+public:
+    ~MessageAuthenticationCode() {}
+
+    /**
+     * @brief Calculates the tag/mac based on the key, the message and the parameters provided to the constructor
+     * @param key The which shall be used for calculation
+     * @param message The message which shall be authenticated
+     * @return The message authentication tag
+     */
+    virtual std::vector<uint8_t> calculate(const std::vector<uint8_t>& key, const std::vector<uint8_t>& message) = 0;
```

Comment:
That interface won't work. Consider the streaming case, where we get the ciphertext in blocks and return it to the user without caching it until finalize is called. We need a similar `update()` and `finish()` based API here.

Luckily, OpenSSL will force you to do this approach anyway in the implementation, so it shouldn't be complicated to just pass this along to the user.